File: backend/app/advisor/scanner.py
"""Market scanner — discovers and ranks top crypto pairs by volume."""

import logging

import ccxt
import pandas as pd

logger = logging.getLogger(__name__)

# Stablecoins and wrapped tokens to exclude from scanning
EXCLUDE_QUOTES = {"USDC", "BUSD", "TUSD", "DAI", "FDUSD", "USDP", "UST"}
EXCLUDE_BASES = {"USDC", "BUSD", "TUSD", "DAI", "FDUSD", "USDP", "UST", "WBTC", "WETH", "STETH"}


class MarketScanner:
    """Scans Binance for top crypto trading pairs by volume."""

    def __init__(self, exchange_id: str = "binance"):
        exchange_class = getattr(ccxt, exchange_id)
        self.exchange = exchange_class({"enableRateLimit": True})

    def scan_top_pairs(self, quote: str = "USDT", top_n: int = 100) -> list[dict]:
        """Fetch all USDT pairs, rank by 24h quote volume, return top N.

        Any pair with < $1M 24h volume is excluded to ensure sufficient
        liquidity for buy/sell orders to fill without slippage issues.

        Returns list of dicts:
            [{symbol, price, volume_24h, change_pct_24h, market_cap_rank}, ...]
        """
        logger.info("Loading markets from %s...", self.exchange.id)
        self.exchange.load_markets()

        # Build set of valid USDT spot pairs, excluding stablecoins
        usdt_symbols = set()
        for symbol, market in self.exchange.markets.items():
            if (
                market.get("quote") == quote
                and market.get("spot", False)
                and market.get("active", True)
                and market.get("base") not in EXCLUDE_BASES
            ):
                usdt_symbols.add(symbol)

        logger.info("Found %d %s spot pairs, fetching tickers...", len(usdt_symbols), quote)

        # Fetch ALL tickers at once (single API call) — avoids Binance
        # symbol-list format issues and is actually faster than passing a list.
        all_tickers = self.exchange.fetch_tickers()

        pairs = []
        for symbol, ticker in all_tickers.items():
            if symbol not in usdt_symbols:
                continue
            quote_volume = ticker.get("quoteVolume") or 0
            if quote_volume < 1_000_000:  # Skip pairs with < $1M daily volume
                continue
            pairs.append({
                "symbol": symbol,
                "price": ticker.get("last") or 0,
                "volume_24h": quote_volume,
                "change_pct_24h": ticker.get("percentage") or 0,
            })

        # Sort by volume descending
        pairs.sort(key=lambda x: x["volume_24h"], reverse=True)

        # Add rank
        for i, p in enumerate(pairs[:top_n]):
            p["rank"] = i + 1

        result = pairs[:top_n]
        logger.info("Top %d pairs by volume (highest: %s at $%.0f vol)",
                     len(result),
                     result[0]["symbol"] if result else "N/A",
                     result[0]["volume_24h"] if result else 0)
        return result

    def fetch_candles_batch(
        self, symbols: list[str], timeframe: str = "1h", limit: int = 200,
    ) -> dict[str, pd.DataFrame]:
        """Fetch candles for multiple symbols.

        Returns dict mapping symbol -> OHLCV DataFrame.
        """
        result = {}
        for i, symbol in enumerate(symbols):
            try:
                ohlcv = self.exchange.fetch_ohlcv(symbol, timeframe, limit=limit)
                if not ohlcv:
                    continue
                df = pd.DataFrame(
                    ohlcv, columns=["time", "open", "high", "low", "close", "volume"],
                )
                df["time"] = pd.to_datetime(df["time"], unit="ms", utc=True)
                result[symbol] = df
                if (i + 1) % 10 == 0:
                    logger.info("Fetched candles for %d/%d symbols...", i + 1, len(symbols))
            except Exception as e:
                logger.warning("Failed to fetch candles for %s: %s", symbol, e)
        logger.info("Fetched candles for %d/%d symbols", len(result), len(symbols))
        return result

File: supabase/functions/_shared/advisor/scanner.ts
/**
 * Market Scanner — discovers and ranks top crypto pairs by volume.
 *
 * Ported from backend/app/advisor/scanner.py
 */

import { getExchange, fetchCandles, type CandleData } from '../ccxt.ts'

// ---------- Constants ----------

const EXCLUDE_BASES = new Set([
  'USDC', 'BUSD', 'TUSD', 'DAI', 'FDUSD', 'USDP', 'UST',
  'WBTC', 'WETH', 'STETH',
])

// ---------- Types ----------

export interface ScannedPair {
  symbol: string
  price: number
  volume_24h: number
  change_pct_24h: number
  rank?: number
}

// ---------- Public API ----------

/**
 * Fetch all USDT pairs, rank by 24h quote volume, return top N.
 *
 * Any pair with < $1M 24h volume is excluded to ensure sufficient
 * liquidity for buy/sell orders to fill without slippage issues.
 */
export async function scanTopPairs(
  exchangeId = 'binance',
  quote = 'USDT',
  topN = 100,
): Promise<ScannedPair[]> {
  const exchange = getExchange(exchangeId)
  console.log(`Loading markets from ${exchangeId}...`)
  await exchange.loadMarkets()

  // Build set of valid USDT spot pairs, excluding stablecoins
  const usdtSymbols = new Set<string>()
  for (const [symbol, market] of Object.entries(exchange.markets)) {
    if (
      market.quote === quote &&
      market.spot === true &&
      market.active !== false &&
      !EXCLUDE_BASES.has(market.base ?? '')
    ) {
      usdtSymbols.add(symbol)
    }
  }

  console.log(`Found ${usdtSymbols.size} ${quote} spot pairs, fetching tickers...`)

  // Fetch ALL tickers at once (single API call)
  const allTickers = await exchange.fetchTickers()

  const pairs: ScannedPair[] = []
  for (const [symbol, ticker] of Object.entries(allTickers)) {
    if (!usdtSymbols.has(symbol)) continue
    const quoteVolume = ticker.quoteVolume ?? 0
    if (quoteVolume < 1_000_000) continue // Skip pairs with < $1M daily volume
    pairs.push({
      symbol,
      price: ticker.last ?? 0,
      volume_24h: quoteVolume,
      change_pct_24h: ticker.percentage ?? 0,
    })
  }

  // Sort by volume descending
  pairs.sort((a, b) => b.volume_24h - a.volume_24h)

  // Add rank
  const result = pairs.slice(0, topN)
  for (let i = 0; i < result.length; i++) {
    result[i].rank = i + 1
  }

  console.log(
    `Top ${result.length} pairs by volume (highest: ${result[0]?.symbol ?? 'N/A'} at $${Math.round(result[0]?.volume_24h ?? 0)} vol)`,
  )
  return result
}

/**
 * Fetch candles for multiple symbols.
 * Returns record mapping symbol -> CandleData[].
 */
export async function fetchCandlesBatch(
  exchangeId: string,
  symbols: string[],
  timeframe = '1h',
  limit = 200,
): Promise<Record<string, CandleData[]>> {
  const result: Record<string, CandleData[]> = {}

  for (let i = 0; i < symbols.length; i++) {
    const symbol = symbols[i]
    try {
      const candles = await fetchCandles(exchangeId, symbol, timeframe, undefined, limit)
      if (candles.length === 0) continue
      result[symbol] = candles
      if ((i + 1) % 10 === 0) {
        console.log(`Fetched candles for ${i + 1}/${symbols.length} symbols...`)
      }
    } catch (e) {
      console.warn(`Failed to fetch candles for ${symbol}: ${e}`)
    }
  }

  console.log(`Fetched candles for ${Object.keys(result).length}/${symbols.length} symbols`)
  return result
}
